### ISSUE TITLE
VM sum query add multiple drop to zero absent in Prometheus

### ISSUE THREAD

**Issue by tbuchier** [VictoriaMetrics/VictoriaMetrics#2505]
Hello,

I'm using vm-cluster v1.75.0, and observing a lot of "drop to zero" when using Victoria-Metrics as datasource that doesn't happens when using Prometheus as datasource.

The query is from a [Mixin Kubernetes](https://github.com/kubernetes-monitoring/kubernetes-mixin) graph.

![image](https://user-images.githubusercontent.com/16001560/165296537-8e95af24-0e97-4f06-9e0b-71b3ef37d50f.png)

![image](https://user-images.githubusercontent.com/16001560/165293323-ecbb9712-37b8-4dda-8ea6-0466f9a00e53.png)

From the dashboard `Kubernetes / Compute Resources / Workload`
Query:
```
sum(
    container_memory_working_set_bytes{cluster="$cluster", namespace="$namespace", container!="", image!=""}
  * on(namespace,pod)
    group_left(workload, workload_type) namespace_workload_pod:kube_pod_owner:relabel{cluster="$cluster", namespace="$namespace", workload="$workload", workload_type="$type"}
) by (pod)
```

I'm facing this behavior in multiple clusters where Prometheus + VM are deployed with 15s scrape interval and latest Kubernetes mixin. Each time is using VM those additional drop to zero are added.

**To Reproduce**
Install VM 1.75
Deploy Kubernetes Mixin

Edit the graph to have a resolution of 1/1 and a min Interval of 15s that correspond to my scraping_interval

I read the https://docs.victoriametrics.com/FAQ.html#why-do-the-same-metrics-have-differences-in-victoriametrics-and-prometheus-dashboards but it didn't mention difference when using a simple sum()

Thanks !

**Comment by YuriKravetc** [comment]
Hi,
Can you attach the troubleshooting graphs for this time period, please?
Also, what happens with resource usage while performing this query?

**Comment by tbuchier** [comment]
Hello,

Here is the troubleshoot graph

![image](https://user-images.githubusercontent.com/16001560/165425153-c4acdef6-097b-4980-825a-63b6f4dd8807.png)


Nothing noticeable happen when running this query, if I select a deployment that only have 1 pod, it's instant as it only fetch container_memory_working_set_bytes for 1 container.


```
resources:
    limits:
      cpu: "12"
      memory: 32Gi
    requests:
      cpu: "8"
      memory: 26Gi
```
![image](https://user-images.githubusercontent.com/16001560/165425555-1c8f6c68-a0b4-4df2-9bef-c5be8e254e89.png)


It's also happening in our other smaller test env (that only ingest 33k dp/sec) and have lower specs

**Comment by valyala** [comment]
@tbuchier , there may be issues with the execution of complex queries. Let's try to narrow down them. Could you run the following query and see whether it also returns frequent zero values:

```promql
sum(container_memory_working_set_bytes{cluster="$cluster", namespace="$namespace", container!="", image!=""}) by (pod)
```

If this query doesn't return frequent zero values, then there is an issue with complex query execution. If the query continues returning frequent zero values, then try executing the following query:

```promql
container_memory_working_set_bytes{cluster="$cluster", namespace="$namespace", container!="", image!=""}
```

If it doesn't return frequent zero values, then the issue is in the `sum()` function. Otherwise there is a high chance that the zero values are actually stored in VictoriaMetrics. This can happen when metrics from multiple scrape targets are stored into the same [time series](https://docs.victoriametrics.com/FAQ.html#what-is-an-active-time-series) because of improper [relabeling](https://docs.victoriametrics.com/vmagent.html#relabeling) configuration. For example, the `container_memory_working_set_bytes` metric could be scraped from `instance1` and `instance2`. Later the relabeling rules could substitute `instance` label for both targets with the same identical value (for example, `instance="foobar"`). In this case both metrics from these targets will be written into the same time series - `container_memory_working_set_bytes{instance="foobar", ...}` . If the metric at one target usually contains values close to zero, while the metric at another target contains bigger values, then the resulting graph will look like in [screenshots above](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2505#issue-1215849377).

**Comment by tbuchier** [comment]
Hello,

Indeed the zero are only there when using sum(). 
When using the raw metrics, I found out that it correspond to period of time where the main containers data have a gap of missing datapoints and the resulting sum is then very small as the other sidecar use only 12.5mb.

![image](https://user-images.githubusercontent.com/16001560/167236897-d0067ef8-7f50-43e3-8f91-535b172f7c74.png)


However from Prometheus the timeserie for this container doesn't have any gap:
![image](https://user-images.githubusercontent.com/16001560/167236933-d6db3844-0174-49a2-9f52-c9468678e0bf.png)



For my remote write config I am using the config mentioned here: https://docs.victoriametrics.com/#prometheus-setup
```
remoteWrite:
      - url: "http://vminsert.... prometheus/api/v1/write"
        queueConfig:
          maxSamplesPerSend: 10000
          capacity: 20000
          maxShards: 30
```

Prom ingestion is constant at 396k/sec, but I don't think it's a load issue as it happens on smaller env with 29k/sec

Pending Datapoints on Busier env:
![image](https://user-images.githubusercontent.com/16001560/167237248-ec351d0d-70ae-45ca-b1c0-480779f79d0c.png)

On Small env:
![image](https://user-images.githubusercontent.com/16001560/167237268-51bc8390-a9d0-4e11-be4e-023eea2b1500.png)

**Comment by tbuchier** [comment]
@valyala  Do you have any suggestion, on what I should check ? 
For now our workaround is either to go back to Prometheus datasource for short term query or to use "null value as connected" when using VictoriaMetrics 

Prom version is v2.35.0

**Comment by willemm** [comment]
We have the same issue (missing data points causing dips in any graph using sum) in v1.85.2  (cluster setup, 4 storage nodes, 2 insert instances, 2 vmagent instances)

How to troubleshoot this?   We're just in the tryout so I could try disbalans disabling vmagent and setting vm as remotewrite target for prometheus?

**Comment by willemm** [comment]
With prometheus as scraper/remotewrite, the same issue seems to occur.

Also, when I look at the graph in question in VMUI, it seems related to the step size.  That is, if I increase the step size sufficiently, the gaps disappear, even though the step size is still well below the scrape interval of the metric.

Is there a way to show the actual data points in VMUI instead of values per step size?

**Comment by hagen1778** [comment]
> However from Prometheus the timeserie for this container doesn't have any gap:
> Also, when I look at the graph in question in VMUI, it seems related to the step size. That is, if I increase the step size sufficiently, the gaps disappear, even though the step size is still well below the scrape interval of the metric.

Looks like both questions are related to staleness detection in VictoriaMetrics:
> If you observe gaps when plotting time series try simplifying your query according to p2 and follow the list. If problem still remains, then it is likely caused by irregular intervals for metrics collection (network delays or targets unavailability on scrapes, irregular pushes, irregular timestamps). VictoriaMetrics automatically [fills the gaps](https://docs.victoriametrics.com/keyconcepts/#range-query) based on median interval between [data samples](https://docs.victoriametrics.com/keyconcepts/#raw-samples). This might work incorrect for irregular data as median will be skewed. In this case it is recommended to switch to the static interval for gaps filling by setting -search.minStalenessInterval=5m cmd-line flag (5m is the static interval used by Prometheus).

from https://docs.victoriametrics.com/troubleshooting/#unexpected-query-results

tl;dr; VictoriaMetrics calculates staleness for each specific time series individually based on median reoslution between samples. It does it so because VictoriaMetrics could be a central database for many telemetry signals, scraped or ingested. Some series could have resolution 10m, some 1s.
While Prometheus uses a static staleness interval of 5m. This means that series with a resolution of 10m will always be plotted with gaps. A series with 1s resolution will produce data points 5m after the last recorded sample.

It is likely series you're plotting don't have stable resolution and either are sporadic or contain duplicates which mess the resolution. You have two options depending on to what extent you want to go:
1. If you don't want to care about data resolution or possible duplicates - go with setting `search.minStalenessInterval=5m` to get similar experience as in Prometheus.
2. If you want to understand your data better - follow https://docs.victoriametrics.com/troubleshooting/#unexpected-query-results


Closing as answered.
